File: hermes/watcher_adapter.py
"""Hermes-side thin adapter for the anti-fixation watcher.

Mirrors openclaw/watcher-adapter.ts but adapts to the Hermes delivery model:
Hermes delivers by EDITING the authoritative response text (the
``transform_llm_output`` hook returns the replacement string), so instead of a
separate gated ``deliver`` call this adapter returns the nudge text to inline
into the prose (the caller composes nudge-into-prose THEN the trailing MEDIA
directive). All gate/branch logic lives here; the hook stays a single composed
path.

Boundaries match the TS adapter: derived-state-only bounded per-scope transcript
store (idle-evicted, LRU-capped), host-owned scope/ids read not minted, nudge
only past an allowing gate, cooldown/dedup committed only when a nudge is
actually inlined. ``shadow_mode`` defaults ON (audit-only). The LLM
critic/generator/moderator are injected; absent on a live path the adapter
fail-closes (no nudge) instead of a permissive default. Targets Python 3.9+.
"""

from __future__ import annotations

from typing import Callable, Dict, List, Optional, Sequence

try:  # works as a package (import hermes.watcher_adapter)
    from . import watcher_core as core
except ImportError:  # works when loaded standalone via spec_from_file_location
    import importlib.util as _ilu
    from pathlib import Path as _Path

    _core_spec = _ilu.spec_from_file_location(
        "hent_ai_watcher_core", _Path(__file__).resolve().parent / "watcher_core.py"
    )
    assert _core_spec is not None and _core_spec.loader is not None
    core = _ilu.module_from_spec(_core_spec)
    _core_spec.loader.exec_module(core)

WATCHER_WINDOW_N = 8
WATCHER_SCOPE_TTL_MS = 1_800_000  # 30 min idle eviction
WATCHER_MAX_SCOPES = 500
DEFAULT_COOLDOWN_MS = 600_000
DEFAULT_BUDGET_PER_HOUR = 20
DEFAULT_CONFIDENCE_THRESHOLD = 0.7
HOUR_MS = 3_600_000

# Injected dependency types (all optional except logger):
#   critic(signal, context, recent_texts) -> Optional[dict{"fixated":bool,"confidence":float}]
#   generate(signal, context) -> Optional[str]
#   moderate(text) -> bool
Critic = Callable[[Dict[str, object], Dict[str, object], Sequence[str]], Optional[Dict[str, object]]]
Generator = Callable[[Dict[str, object], Dict[str, object]], Optional[str]]
Moderator = Callable[[str], bool]


def _now_ms() -> int:
    import time

    return int(time.time() * 1000)


def create_hermes_watcher_adapter(deps: Dict[str, object]):
    cfg: Dict[str, object] = deps.get("config") or {}
    logger = deps["logger"]
    shadow_mode = bool(cfg.get("shadowMode", True))
    cooldown_ms = int(cfg.get("cooldownMs", DEFAULT_COOLDOWN_MS))
    budget_per_hour = int(cfg.get("budgetPerHour", DEFAULT_BUDGET_PER_HOUR))
    confidence_threshold = float(cfg.get("confidenceThreshold", DEFAULT_CONFIDENCE_THRESHOLD))
    enabled = bool(cfg.get("enabled", False))
    critic: Optional[Critic] = deps.get("critic")  # type: ignore[assignment]
    generate: Optional[Generator] = deps.get("generate")  # type: ignore[assignment]
    moderate: Optional[Moderator] = deps.get("moderate")  # type: ignore[assignment]
    now: Callable[[], int] = deps.get("now") or _now_ms  # type: ignore[assignment]
    iso_now: Callable[[], str] = deps.get("isoNow") or (lambda: core.DEFAULT_NOW)  # type: ignore[assignment]

    buffers: "Dict[str, Dict[str, object]]" = {}
    budgets: "Dict[str, Dict[str, int]]" = {}
    last_delivered: Dict[str, int] = {}
    delivered_signals = set()
    state = {"seq": 0}

    def evict_idle(current: int) -> None:
        for scope_id in [s for s, b in buffers.items() if current - int(b["lastTouched"]) > WATCHER_SCOPE_TTL_MS]:
            del buffers[scope_id]

    def push(scope_id: str, message: Dict[str, object]) -> List[Dict[str, object]]:
        current = now()
        evict_idle(current)
        buffer = buffers.pop(scope_id, None)
        if buffer is None:
            buffer = {"messages": [], "lastTouched": current}
        msgs: List[Dict[str, object]] = buffer["messages"]  # type: ignore[assignment]
        msgs.append(message)
        if len(msgs) > WATCHER_WINDOW_N:
            buffer["messages"] = msgs[-WATCHER_WINDOW_N:]
        buffer["lastTouched"] = current
        buffers[scope_id] = buffer  # re-insert at tail (LRU order)
        while len(buffers) > WATCHER_MAX_SCOPES:
            del buffers[next(iter(buffers))]
        return buffer["messages"]  # type: ignore[return-value]

    def within_budget(scope_id: str, current: int) -> bool:
        budget = budgets.get(scope_id)
        if budget is None or current - budget["windowStart"] >= HOUR_MS:
            budget = {"windowStart": current, "count": 0}
            budgets[scope_id] = budget
        if budget["count"] >= budget_per_hour:
            return False
        budget["count"] += 1
        return True

    def record_agent_turn(scope_id: str, text: str) -> List[Dict[str, object]]:
        state["seq"] += 1
        return push(scope_id, {"id": f"h-{state['seq']}", "senderRole": "agent", "ts": iso_now(), "text": text})

    def run_live(signal, messages, scope_id, cooldown_key, current) -> Optional[str]:
        if not critic or not generate or not moderate:
            logger.warn("watcher: live mode but LLM critic/generator/moderator missing, fail-closed")
            return None
        if not within_budget(scope_id, current):
            logger.warn(f"watcher: critic budget exceeded for scope={scope_id}, fail-closed (no nudge)")
            return None
        try:
            context = core.create_neutral_conversation_context(scope_id, messages)
            verdict = critic(signal, context, [str(m.get("text", "")) for m in messages])
            if not verdict:
                logger.warn("watcher: critic returned null, fail-closed (no nudge)")
                return None
            if not verdict.get("fixated") or float(verdict.get("confidence", 0)) < confidence_threshold:
                return None
            text = generate(signal, context)
            if not text or not moderate(text):
                logger.warn("watcher: nudge suppressed (empty generation or moderation failed)")
                return None
        except Exception:  # a misbehaving LLM dep must never crash delivery
            logger.warn("watcher: LLM pipeline raised, fail-closed (no nudge)")
            return None
        last_delivered[cooldown_key] = current
        delivered_signals.add(f"{scope_id}:{signal['signalId']}")
        return text

    def on_agent_turn(scope_id: str, text: str) -> Optional[str]:
        """Record the agent turn; return nudge text to inline, or None."""
        messages = record_agent_turn(scope_id, text)
        if not enabled:
            return None
        signals = core.evaluate_fixation(messages, scope_id)
        if not signals:
            return None
        signal = signals[0]
        cooldown_key = f"{scope_id}:{signal['fixationPattern']}"
        current = now()
        cooldown_hit = current - last_delivered.get(cooldown_key, -(2 ** 63)) < cooldown_ms
        duplicate_hit = f"{scope_id}:{signal['signalId']}" in delivered_signals

        nudge_text: Optional[str] = None
        if not shadow_mode and not cooldown_hit and not duplicate_hit:
            nudge_text = run_live(signal, messages, scope_id, cooldown_key, current)

        audit = core.evaluate_host_policy_gate(
            {
                "runtime": "hermes",
                "signal": signal,
                "criticConfidence": 0,
                "shadowMode": shadow_mode,
                "cooldownHit": cooldown_hit,
                "duplicateHit": duplicate_hit,
                "deliveryMessageId": f"inline-{scope_id}" if nudge_text else None,
                "now": iso_now(),
            }
        )
        logger.info(
            f"watcher: scope={scope_id} pattern={signal['fixationPattern']} allowed={audit['allowed']}"
            f" nudged={'yes' if nudge_text else 'no'}"
            + (f" suppressed={audit['suppressedReason']}" if audit.get("suppressedReason") else "")
        )
        return nudge_text

    return {
        "on_agent_turn": on_agent_turn,
        "scope_count": lambda: len(buffers),
    }


def compose_nudge(response_text: str, nudge_text: str) -> str:
    """Replace the would-be repeat with the agent-authored nudge (identity: agent_explicit).

    Mode B (replace): when a stale self-repetition fires, the duplicate response
    prose is dropped entirely and ONLY the steer is emitted (the emotion MEDIA
    directive is still appended downstream by build_transformed_response). The
    ``response_text`` argument is retained for call-site symmetry.
    """
    return nudge_text.strip()

File: openclaw/watcher-adapter.ts
// Thin OpenClaw adapter for the anti-fixation watcher. Holds ALL gate/branch
// logic so openclaw/index.ts stays a single thin delegation per hook (keeps the
// large index.ts out of new branch coverage). Pure core lives in watcher-core.ts.
//
// Boundaries (see the approved plan): derived-state-only (a bounded, idle-evicted
// per-scope turn buffer reconstructable from host events), host-owned ids read
// not minted, deliver only past an allowing gate, and cooldown/dedup committed
// ONLY on a real host delivery id (a dropped send stays retryable). shadow_mode
// defaults ON so v1 is audit-only until explicitly promoted to live.
//
// The LLM critic / generator / moderator are INJECTED (real implementations land
// with the LLM layer). When they are absent on a live path the adapter
// fail-closes (no nudge) rather than masking with a permissive default.

import {
  createNeutralConversationContext,
  evaluateFixation,
  evaluateHostPolicyGate,
  type HostPolicyGateAudit,
  type InternalAntiFixationSignal,
  type NeutralConversationContext,
  type RawConversationMessage,
} from "./watcher-core.js";

export interface WatcherConfig {
  /** Opt-in; the watcher does nothing unless enabled. Default false. */
  enabled?: boolean;
  /** Evaluate + audit but never deliver. Default true. */
  shadowMode?: boolean;
  /** Min ms between deliveries for the same scope:fixationPattern. Default 600000. */
  cooldownMs?: number;
  /** Max critic calls per channel per hour. Default 20. */
  budgetPerHour?: number;
  /** Min critic confidence to deliver. Default 0.7. */
  confidenceThreshold?: number;
}

export interface WatcherLogger {
  info: (...args: unknown[]) => void;
  warn: (...args: unknown[]) => void;
}

export interface CriticResult {
  fixated: boolean;
  confidence: number;
}
export type WatcherCritic = (input: {
  signal: InternalAntiFixationSignal;
  context: NeutralConversationContext;
  recentTexts: string[];
}) => Promise<CriticResult | null>;
export type WatcherGenerator = (input: {
  signal: InternalAntiFixationSignal;
  context: NeutralConversationContext;
}) => Promise<string | null>;
export type WatcherModerator = (text: string) => boolean;
/** Host-native delivery: returns the host-assigned delivery id, or null if dropped. */
export interface WatcherDeliverOptions {
  /**
   * Host message id the steer should REPLACE in place (Mode B). The host decides
   * how: OpenClaw edits that Discord message; a host without an edit capability
   * falls back to posting the steer as a fresh message.
   */
  replaceMessageId?: string;
}
export type WatcherDeliver = (
  channelId: string,
  text: string,
  opts?: WatcherDeliverOptions,
) => Promise<string | null>;

export interface WatcherAdapterDeps {
  config?: WatcherConfig;
  logger: WatcherLogger;
  deliver: WatcherDeliver;
  critic?: WatcherCritic;
  generate?: WatcherGenerator;
  moderate?: WatcherModerator;
  /** Monotonic clock (ms) for cooldown/budget/eviction. Default Date.now. */
  now?: () => number;
  /** ISO clock for audit timestamps. Default new Date().toISOString. */
  isoNow?: () => string;
}

export const WATCHER_WINDOW_N = 8;
export const WATCHER_SCOPE_TTL_MS = 1_800_000; // 30 min idle eviction
export const WATCHER_MAX_SCOPES = 500;
const DEFAULT_COOLDOWN_MS = 600_000;
const DEFAULT_BUDGET_PER_HOUR = 20;
const DEFAULT_CONFIDENCE_THRESHOLD = 0.7;
const HOUR_MS = 3_600_000;

interface ScopeBuffer {
  messages: RawConversationMessage[];
  lastTouched: number;
}

interface ChannelBudget {
  windowStart: number;
  count: number;
}

export interface OnAgentTurnArgs {
  scopeId: string;
  channelId: string;
  text: string;
  messageId: string;
  sourceThreadId?: string;
  targetThreadId?: string;
  sessionId?: string;
}

export interface OpenClawWatcherAdapter {
  recordUserTurn(scopeId: string, text: string, id?: string): void;
  onAgentTurn(args: OnAgentTurnArgs): Promise<HostPolicyGateAudit | null>;
  /** Number of live scope buffers (after eviction); for diagnostics/tests. */
  scopeCount(): number;
}

export function createOpenClawWatcherAdapter(deps: WatcherAdapterDeps): OpenClawWatcherAdapter {
  const cfg = deps.config ?? {};
  const shadowMode = cfg.shadowMode ?? true;
  const cooldownMs = cfg.cooldownMs ?? DEFAULT_COOLDOWN_MS;
  const budgetPerHour = cfg.budgetPerHour ?? DEFAULT_BUDGET_PER_HOUR;
  const confidenceThreshold = cfg.confidenceThreshold ?? DEFAULT_CONFIDENCE_THRESHOLD;
  const now = deps.now ?? (() => Date.now());
  const isoNow = deps.isoNow ?? (() => new Date().toISOString());

  const buffers = new Map<string, ScopeBuffer>();
  const budgets = new Map<string, ChannelBudget>();
  const lastDelivered = new Map<string, number>(); // cooldownKey -> ms
  const deliveredSignals = new Set<string>(); // signalId
  let seq = 0;

  function evictIdle(current: number): void {
    for (const [scopeId, buffer] of buffers) {
      if (current - buffer.lastTouched > WATCHER_SCOPE_TTL_MS) buffers.delete(scopeId);
    }
  }

  function push(scopeId: string, message: RawConversationMessage): RawConversationMessage[] {
    const current = now();
    evictIdle(current);
    let buffer = buffers.get(scopeId);
    if (buffer) {
      buffers.delete(scopeId); // re-insert at the tail so iteration order is LRU
    } else {
      buffer = { messages: [], lastTouched: current };
    }
    buffer.messages.push(message);
    if (buffer.messages.length > WATCHER_WINDOW_N) {
      buffer.messages = buffer.messages.slice(-WATCHER_WINDOW_N);
    }
    buffer.lastTouched = current;
    buffers.set(scopeId, buffer);
    while (buffers.size > WATCHER_MAX_SCOPES) {
      const lru = buffers.keys().next().value as string;
      buffers.delete(lru);
    }
    return buffer.messages;
  }

  function withinBudget(channelId: string, current: number): boolean {
    let budget = budgets.get(channelId);
    if (!budget || current - budget.windowStart >= HOUR_MS) {
      budget = { windowStart: current, count: 0 };
      budgets.set(channelId, budget);
    }
    if (budget.count >= budgetPerHour) return false;
    budget.count += 1;
    return true;
  }

  function recordUserTurn(scopeId: string, text: string, id?: string): void {
    seq += 1;
    push(scopeId, { id: id ?? `u-${seq}`, senderRole: "user", ts: isoNow(), text });
  }

  async function runLivePipeline(
    args: OnAgentTurnArgs,
    signal: InternalAntiFixationSignal,
    messages: RawConversationMessage[],
    cooldownKey: string,
    current: number,
  ): Promise<{ deliveryMessageId?: string; criticConfidence: number }> {
    if (!deps.critic || !deps.generate || !deps.moderate) {
      deps.logger.warn("watcher: live mode but LLM critic/generator/moderator missing, fail-closed");
      return { criticConfidence: 0 };
    }
    if (!withinBudget(args.channelId, current)) {
      deps.logger.warn(
        `watcher: critic budget exceeded for channel=${args.channelId}, fail-closed (no nudge)`,
      );
      return { criticConfidence: 0 };
    }
    try {
      const context = createNeutralConversationContext(args.scopeId, messages);
      const verdict = await deps.critic({ signal, context, recentTexts: messages.map((m) => m.text) });
      if (!verdict) {
        deps.logger.warn("watcher: critic returned null, fail-closed (no nudge)");
        return { criticConfidence: 0 };
      }
      if (!verdict.fixated || verdict.confidence < confidenceThreshold) {
        return { criticConfidence: verdict.confidence };
      }
      const text = await deps.generate({ signal, context });
      if (!text || !deps.moderate(text)) {
        deps.logger.warn("watcher: nudge suppressed (empty generation or moderation failed)");
        return { criticConfidence: verdict.confidence };
      }
      const deliveryMessageId = await deps.deliver(args.channelId, text, {
        replaceMessageId: args.messageId,
      });
      if (deliveryMessageId) {
        lastDelivered.set(cooldownKey, current);
        deliveredSignals.add(`${args.scopeId}:${signal.signalId}`);
      }
      return { deliveryMessageId: deliveryMessageId ?? undefined, criticConfidence: verdict.confidence };
    } catch {
      deps.logger.warn("watcher: LLM pipeline raised, fail-closed (no nudge)");
      return { criticConfidence: 0 };
    }
  }

  async function onAgentTurn(args: OnAgentTurnArgs): Promise<HostPolicyGateAudit | null> {
    const messages = push(args.scopeId, {
      id: args.messageId,
      senderRole: "agent",
      ts: isoNow(),
      text: args.text,
      threadId: args.sourceThreadId,
      sessionId: args.sessionId,
    });
    if (!cfg.enabled) return null;

    const signal = evaluateFixation(messages, args.scopeId)[0];
    if (!signal) return null;

    const cooldownKey = `${args.scopeId}:${signal.fixationPattern}`;
    const current = now();
    const cooldownHit = current - (lastDelivered.get(cooldownKey) ?? Number.NEGATIVE_INFINITY) < cooldownMs;
    const duplicateHit = deliveredSignals.has(`${args.scopeId}:${signal.signalId}`);

    let deliveryMessageId: string | undefined;
    let criticConfidence = 0;
    if (!shadowMode && !cooldownHit && !duplicateHit) {
      const outcome = await runLivePipeline(args, signal, messages, cooldownKey, current);
      deliveryMessageId = outcome.deliveryMessageId;
      criticConfidence = outcome.criticConfidence;
    }

    const audit = evaluateHostPolicyGate({
      runtime: "openclaw",
      signal,
      criticConfidence,
      sourceThreadId: args.sourceThreadId,
      targetThreadId: args.targetThreadId,
      sessionId: args.sessionId,
      shadowMode,
      cooldownHit,
      duplicateHit,
      deliveryMessageId,
      now: isoNow(),
    });
    deps.logger.info(
      `watcher: scope=${args.scopeId} pattern=${signal.fixationPattern} allowed=${audit.allowed}` +
        ` delivered=${deliveryMessageId ? "yes" : "no"}` +
        `${audit.suppressedReason ? ` suppressed=${audit.suppressedReason}` : ""}`,
    );
    return audit;
  }

  return { recordUserTurn, onAgentTurn, scopeCount: () => buffers.size };
}
